src/snowcli/cli/stage/diff.py

```diff
@@ -0,0 +1,151 @@
+    "Files with matching md5sums"
+
+    different: List[str] = field(default_factory=list)
+    "Files that may be different between the stage and the local directory"
+
+    only_local: List[str] = field(default_factory=list)
+    "Files that only exist in the local directory"
+
+    only_on_stage: List[str] = field(default_factory=list)
+    "Files that only exist on the stage"
+
+    def display_pretty_diff(self):
+        return ObjectResult(
+            {
+                "only on local": ", ".join(self.only_local),
+                "only on stage": ", ".join(self.only_on_stage),
+                "modified/unknown": ", ".join(self.different),
+                "identical": ", ".join(self.identical),
+            }
+        )
+
+
+def is_valid_md5sum(checksum: str) -> bool:
+    """
+    Could the provided hexadecimal checksum represent a valid md5sum?
+    """
+    return re.match(MD5SUM_REGEX, checksum) is not None
+
+
+def compute_md5sum(file: Path) -> str:
+    """
+    Returns a hexidecimal checksum for the file located at the given path.
+    """
+    if not file.is_file():
+        raise ValueError(
+            "The provided file does not exist or not a (symlink to a) regular file"
+        )
+
+    # FIXME: there are two cases in which this will fail to provide a matching
+    # md5sum, even when the underlying file is the same:
+    #  1. when the stage uses SNOWFLAKE_FULL encryption
+    #  2. when the file was uploaded in multiple parts
+
+    # We can re-create the second if we know what chunk size was used by the
+    # upload process to the backing object store (e.g. S3, azure blob, etc.)
+    # but we cannot re-create the first as the encrpytion key is hidden.
+
+    # We are assuming that we will not get accidental collisions here due to the
+    # large space of the md5sum (32 * 4 = 128 bits means 1-in-9-trillion chance)
+    # combined with the fact that the file name + path must also match elsewhere.
+
+    with open(file, "rb") as f:
+        file_hash = hashlib.md5()
+        while chunk := f.read(CHUNK_SIZE_BYTES):
+            file_hash.update(chunk)
+
+    return file_hash.hexdigest()
+
+
+def enumerate_files(path: Path) -> List[Path]:
+    """
+    Get a list of all files in a directory (recursively).
+    """
+    if not path.is_dir():
+        raise ValueError("Path must point to a directory")
+
+    paths: List[Path] = []
+    for child in sorted(path.iterdir()):
+        if child.is_dir():
+            paths += enumerate_files(child)
+        else:
+            paths.append(child)
+
+    return paths
+
+
+def strip_stage_name(path: str) -> str:
+    """Returns the given stage path without the stage name as the first part."""
+    return "/".join(path.split("/")[1:])
+
+
+def build_md5_map(list_stage_cursor: SnowflakeCursor) -> Dict[str, str]:
+    """
+    Returns a mapping of relative stage paths to their md5sums.
+    """
+    return {
+        # XXX: how can we get dicts back here?
```

Comment:
What about this comment?
Is it TODO?
Is it about sth what is required to have everything working correctly?

Reply:
I'd delete this comment if we can't understand it.

IMO it's introducing only some confusion.